### PR TITLE
[primary] Aggressive pruning of old message cancel handles 

### PR DESCRIPTION
**Pull request by gdanezis**
Once the primary receives a valid certificate for round X, there is no more point to wait for messages that were received in rounds lower than X to advance rounds and the DAG. Any missing information ging forward might be re-requested under a higher round number, or by other subsystems like synchronizer that only gets messages GCed after a commit.

### CODE REVIEW

**Comment by gdanezis** [comment]
hey @asonnino any idea why the reconfig tests take for ever with this change?
```
        SLOW [> 60.001s]                        primary::epoch_change test_restart_with_new_committee_change
        SLOW [> 60.001s]                            node::reconfigure restart
```

File: primary/src/core.rs [hunk 1]
```diff
@@ -423,6 +423,11 @@ impl Core {
                 .send((parents, certificate.round(), certificate.epoch()))
                 .await
                 .map_err(|_| DagError::ShuttingDown)?;
+            
+            debug!("Cancel Handlers len: {}", self.cancel_handlers.len());
+            if certificate.round() > 0 {
+                self.cancel_handlers.retain(|k, _| *k > certificate.round()-1);
+            }
```
Comment (huitseeker) [inline primary/src/core.rs:430]:
```suggestion
            let before = self.cancel_handlers.len();
            if certificate.round() > 0 {
                self.cancel_handlers.retain(|k, _| *k >= certificate.round());
            }
            debug!("Pruned {} messages from obsolete rounds.", self.cancel_handlers.len() - before);
```

**Review by huitseeker — COMMENTED** [review]
You cancel the following possible messages with this:
1. broadcasting a new header,
2. broadcasting a new vote,
3. broadcasting a new certificate.

A very important note: you're not cancelling *individual, point to point messages* but rather canceling the whole broadcast itself, in all its instances. Which concerns me for (3.)

I suspect that this is likely, as early as the first certificate of round n, to prevent the arrival of a k-th certificate at round (n-1), with 2f+1 < k < 3f+1. Couldn't that certificate still plausibly be used as a parent in one of the 2d -> 2f+1-th blocks of round n?

(I see no objection with `self.cancel_handlers.retain(|k, _| *k >= certificate.round() - 1)` though).

**Comment by asonnino** [comment]
> hey @asonnino any idea why the reconfig tests take for ever with this change?
> 
> ```
>         SLOW [> 60.001s]                        primary::epoch_change test_restart_with_new_committee_change
>         SLOW [> 60.001s]                            node::reconfigure restart
> ```

No that's not normal, let me check what may have gone wrong

**Comment by asonnino** [comment]
I would also go for `self.cancel_handlers.retain(|k, _| *k >= certificate.round() - 1)`. When running the DAG in partial synchrony (ie. to run Bullshark), a Byzantine node may not wait for the leader of round R and immediately produce a certificate for round R+1. Then, the honest leader of round R may receive the certificate of round R+1 and stop sending his own certificate of round R. At this point all honest nodes will hit the 5sec timeout before advancing to the next DAG round. 

I am not sure I understand @huitseeker's argument when running the DAG in asynchronous (ie. for Tusk).

File: benchmark/benchmark/logs.py [hunk 2]
```diff
@@ -158,7 +158,7 @@ def _parse_workers(self, log):
         return sizes, samples, ip
 
     def _to_posix(self, string):
-        x = parser.isoparse(string)
+        x = parser.isoparse(string[:24])
```
[the remark this answers is not in the document]
Comment (gdanezis) [reply benchmark/benchmark/logs.py:161]: In my version of Python the ISO formater is strict and only accpets up to 6 digits after the dot. Whereas our rust logging lib prints more :(

File: primary/src/core.rs [hunk 3]
```diff
@@ -423,6 +423,11 @@ impl Core {
                 .send((parents, certificate.round(), certificate.epoch()))
                 .await
                 .map_err(|_| DagError::ShuttingDown)?;
+            
+            debug!("Cancel Handlers len: {}", self.cancel_handlers.len());
+            if certificate.round() > 0 {
```
[the remark this answers is not in the document]
Comment (gdanezis) [reply primary/src/core.rs:428]: I just wanted to be trivially correct here an not have a negative round

**Comment by gdanezis** [comment]
> > hey @asonnino any idea why the reconfig tests take for ever with this change?
> > ```
> >         SLOW [> 60.001s]                        primary::epoch_change test_restart_with_new_committee_change
> >         SLOW [> 60.001s]                            node::reconfigure restart
> > ```
> 
> No that's not normal, let me check what may have gone wrong (the test is supposed to take about 8s)

Hm this may also be due to the base PR that does the channel stats. Will rebase on main.

**Comment by asonnino** [comment]
> hey @asonnino any idea why the reconfig tests take for ever with this change?
> 
> ```
>         SLOW [> 60.001s]                        primary::epoch_change test_restart_with_new_committee_change
>         SLOW [> 60.001s]                            node::reconfigure restart
> ```

This is very strange. The latest main runs fine but the problem shows on this PR only (even after rebasing on main)

**Comment by asonnino** [comment]
> > hey @asonnino any idea why the reconfig tests take for ever with this change?
> > ```
> >         SLOW [> 60.001s]                        primary::epoch_change test_restart_with_new_committee_change
> >         SLOW [> 60.001s]                            node::reconfigure restart
> > ```
> 
> This is very strange. The latest main runs fine but the problem shows on this PR only (even after rebasing on main)

Ok I think the issue is not directly in this PR, but in this commit: https://github.com/MystenLabs/narwhal/commit/71be4bb8fbbf2a5d233d47b4c19bffc48d7436f5

There is a task in Primary::spawn (line 127) that never exits upon reboot.

**Comment by gdanezis** [comment]
>To repeat what I said, at retain(|k| *k >= certificate.round() -1) this gets an insta-approve.

I agree this is a little more conservative and still gets us most of the benefit!

File: primary/src/core.rs [hunk 4]
```diff
@@ -423,6 +423,16 @@ impl Core {
                 .send((parents, certificate.round(), certificate.epoch()))
                 .await
                 .map_err(|_| DagError::ShuttingDown)?;
+
+            let before = self.cancel_handlers.len();
+            if certificate.round() > 0 {
+                self.cancel_handlers
```
Comment (huitseeker) [inline primary/src/core.rs:429]: Nit: if we want to track the debate on whether this bound of `certificate.round()` can be fine-tuned `certificate.round()-1`, a comment here or an issue would do nicely.

**Review by huitseeker — APPROVED** [review]
Thank you @gdanezis !

**Pull request merged**